### PR TITLE
Ct int configuration

### PR DESCRIPTION
**Pull request by gbalduzz**
First PR for merging CT-INT. Merges the configuration and matrix configuration objects.

### CODE REVIEW

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+                                               std::vector<int>& from, std::vector<int>& to) {
+  for (int s = 0; s < 2; ++s) {
+    // Sort and prepare source array.
+    auto& sector = BaseClass::sectors_[s].entries_;
+    std::sort(sector_to[s].begin(), sector_to[s].end());
+    auto& tags = BaseClass::sectors_[s].tags_;
+    sector_from[s].clear();
+    int source_idx = sector.size() - sector_to[s].size();
+    for (int i = 0; source_idx < sector.size(); ++i, ++source_idx) {
+      while (std::binary_search(sector_to[s].begin(), sector_to[s].end(), source_idx))
```
Comment (PDoakORNL) [inline include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: I found these dead i indexes, while loops and checks for sector size overun very confusion. By the time I figured out they were correct (I think) i had figured out a significantly easier way to write them.  See following PR. I pass the test so if my PR is wrong then the testing is insufficient.

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+  std::vector<int> randomRemovalCandidate(RngType& rng, double removal_rand);
+  template <class RngType>
+  std::vector<int> randomRemovalCandidate(RngType& rng);
+
+  // Remove elements with id remove by copying elements from the end.
+  // Postcondition: from and sector_from contains the indices that where moved into the place
+  //                of the old elements.
+  // Postcondition: all vectors are ordered, resized, and ready to be used for moving the rows of
+  //                the M matrix
+  // In/Out: sector_remove, remove.
```
Comment (PDoakORNL) [inline include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: These are probably better variable names. But change them here or in the declaration they need to match.

File: include/dca/linalg/util/allocators/pinned_allocator.hpp [hunk 3]
```diff
@@ -49,6 +56,13 @@ class PinnedAllocator : public std::allocator<T> {
     }
     ptr = nullptr;
   }
+
+  constexpr bool operator==(const PinnedAllocator<T>& /*other*/) {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/linalg/util/allocators/pinned_allocator.hpp]: OK, I merged the PR.

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_matrix_configuration.hpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    const InteractionElement& elem = (*H_int_)[v.interaction_id];
+    if (elem.partners_id.size())
+      return leg == 1 ? -3 * sign : 3 * sign;  // non density-density.
+    else if (elem.w > 0)
+      return leg == 1 ? -1 * sign : 1 * sign;  // positive dd interaction.
+    else
+      return 2 * sign;  // negative dd interaction.
+  };
+
+  std::array<int, 2> sizes{0, 0};
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_matrix_configuration.hpp]: With multiband interactions, you can have same spin interactions, e.g. n(up, 0) n(down, 1). So the sizes change for a single vertex update can be {0, 2} {1, 1} or {2, 0}.

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/utils.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace dca {
+namespace phys {
+namespace solver {
+namespace ctint {
+namespace details {
+// dca::phys::solver::ctint::details::
+
+template<class T>
+T getRandomElement(const std::vector<std::vector<T>*>& v_ptrs, const double rand){
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/utils.hpp]: It does not own it.  What is the advantage of using it compared to a raw pointer? Isn't the syntax vctrs[i]->size() more clear than  vctrs[i].get().size()?

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp (hunk 2 above)
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: Changed name of  the arguments to match the documentation in 35b02bc5a9f92c7d3a74ad038a94b299ee0f3fe1

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp (hunk 1 above)
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: Removed index i in 35b02bc5a9f92c7d3a74ad038a94b299ee0f3fe1

File: test/unit/phys/dca_step/cluster_solver/ctint/structs/solver_configuration_test.cpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+  EXPECT_EQ(3, config.getSector(0).size());
+  EXPECT_EQ(1, config.getSector(1).size());
+
+  // See above test regarding the consistency of the configuration.
+  // TODO: always leave config consitent.
+  EXPECT_FALSE(config.checkConsistency());
+}
+
+TEST(SolverConfigurationTest, ShrinkAndMove) {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply test/unit/phys/dca_step/cluster_solver/ctint/structs/solver_configuration_test.cpp]: Fixed in 35b02bc5a9f92c7d3a74ad038a94b299ee0f3fe1

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_sector.hpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+  auto getTag(int i) const {
+    return tags_[i];
+  }
+
+  bool operator==(const Sector& rhs) const {
+    return entries_ == rhs.entries_;
+  }
+
+protected:
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_sector.hpp]: Forgot I ended up not inheriting from it. Fixed in 35b02bc5a9f92c7d3a74ad038a94b299ee0f3fe1

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/utils.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/utils.hpp]: Clear ownership semantics.  Standard unique insures destruction.  Using reference_wrapper makes it clear it is not owned. If the compiler can deduce whether you want the reference or the reference_wrapper you don't need get().

example:
```C++
template<typename T>
using RefVector<T> = std::vector<std::reference_wrapper<T>>>;
...
RefVector<std::vector<int>> v_refs;
...
for(std::vector<int>& v_ref : v_refs)
        size += v_ref.size();
```

Anyway think about it, it is a stylistic change.

**Pull request merged**